harness/determined/cli/project.py

```diff
@@ -0,0 +1,303 @@
+import json
+from argparse import Namespace
+from typing import Any, Dict, List, Sequence, Tuple
+
+from determined.cli.session import setup_session
+from determined.common.api import authentication, bindings
+from determined.common.api.bindings import v1Experiment, v1Project, v1Workspace
+from determined.common.declarative_argparse import Arg, Cmd
+from determined.common.experimental import session
+
+from . import render
+from .workspace import list_workspace_projects, workspace_by_name
+
+
+def render_experiments(args: Namespace, experiments: Sequence[v1Experiment]) -> None:
+    def format_experiment(e: v1Experiment) -> List[Any]:
+        result = [
+            e.id,
+            e.username,
+            e.name,
+            e.forkedFrom,
+            e.state.value.replace("STATE_", ""),
+            render.format_percent(e.progress),
+            render.format_time(e.startTime),
+            render.format_time(e.endTime),
+            e.resourcePool,
+        ]
+        if args.all:
+            result.append(e.archived)
+        return result
+
+    headers = [
+        "ID",
+        "Owner",
+        "Name",
+        "Parent ID",
+        "State",
+        "Progress",
+        "Start Time",
+        "End Time",
+        "Resource Pool",
+    ]
+    if args.all:
+        headers.append("Archived")
+    values = [format_experiment(e) for e in experiments]
+    render.tabulate_or_csv(headers, values, False)
+
+
+def render_project(project: v1Project) -> None:
+    values = [
+        project.id,
+        project.name,
+        project.description,
+        project.numExperiments,
+        project.numActiveExperiments,
+    ]
+    PROJECT_HEADERS = ["ID", "Name", "Description", "# Experiments", "# Active Experiments"]
+    render.tabulate_or_csv(PROJECT_HEADERS, [values], False)
+
+
+def project_by_name(
+    sess: session.Session, workspace_name: str, project_name: str
+) -> Tuple[v1Workspace, v1Project]:
+    w = workspace_by_name(sess, workspace_name)
+    p = bindings.get_GetWorkspaceProjects(sess, id=w.id, name=project_name).projects
+    if len(p) == 0:
+        raise Exception(f'No project found on this workspace with name: "{project_name}"')
```

Comment:
✅

Reply:
ok I had said checkmark but actually the api.errors.NotFoundException is based on a `requests.Response` not an error string. Generating a Response object is possible here but not a good fit - as this is CLI code running locally. We do a good job of returning an appropriate not-found server error in the Go code

```
	case db.ErrNotFound:
		return nil, status.Errorf(
			codes.NotFound, "project (%d) not found", id)
```

Reply:
ok, then maybe add a new `class EmptyResultException(BadResponseException):`?

Reply:
I'd still prefer a typed exception hierarchy.

Reply:
Should I be handling CLI exceptions more like we do in Experiments?
```
    except Exception as e:
        print("Failed to initialize git repository at ", "{}: {}".format(repo_path, e))
        sys.exit(1)
```

Reply:
I'd say throwing an exception should be enough. calling `sys.exit` in the middle of the code is a bad pattern in my opinion, because you can't effectively reuse these functions elsewhere then.

Reply:
ok here's my take on `EmptyResultException` and `ForbiddenException` in the areas where we are having this part of the CLI raise exceptions.  If this looks good, I'll make a ticket to apply these to the Models CLI at least.

Reply:
Link: https://github.com/determined-ai/determined/pull/4203/commits/8258a165e5e2dbf053745ce93f0796dc62f60bf5